packages/medusa-payment-stripe/src/api/hooks/stripe.ts

```diff
@@ -0,0 +1,161 @@
+import { Request, Response } from "express"
+import {
+  buildHandleCartPaymentErrorMessage,
+  constructWebhook,
+  isPaymentCollection,
+} from "../utils/utils"
+import {
+  AbstractCartCompletionStrategy,
+  CartService,
+  IdempotencyKeyService,
+} from "@medusajs/medusa"
+import { MedusaError } from "medusa-core-utils"
+
+export default async (req: Request, res: Response) => {
+  let event
+  try {
+    event = constructWebhook({
+      signature: req.headers["stripe-signature"],
+      body: req.body,
+      container: req.scope,
+    })
+  } catch (err) {
+    res.status(400).send(`Webhook Error: ${err.message}`)
+    return
+  }
+
+  const paymentIntent = event.data.object
+  const cartId = paymentIntent.metadata.cart_id // Backward compatibility
+  const resourceId = paymentIntent.metadata.resource_id
+
+  if (isPaymentCollection(resourceId)) {
+    await handlePaymentCollection(event, req, res, resourceId, paymentIntent.id)
+  } else {
+    await handleCartPayments(event, req, res, cartId ?? resourceId)
+  }
+}
+
+async function handleCartPayments(event, req, res, cartId) {
+  const manager = req.scope.resolve("manager")
+  const orderService = req.scope.resolve("orderService")
+  const logger = req.scope.resolve("logger")
+
+  const order = await orderService
+    .retrieveByCartId(cartId)
+    .catch(() => undefined)
+
+  // handle payment intent events
+  switch (event.type) {
+    case "payment_intent.succeeded":
```

Comment:
thought(non-blocking): As discussed [here](https://github.com/medusajs/medusa/pull/3257#issuecomment-1445333454), we should consider adding order creation to this step to cover the case of customers loosing their connection or leaving the completion-page before the full cart-completion flow has been run.